### PR TITLE
Gh 508 mnemonic to pk

### PR DESCRIPTION
**Pull request by sche**

### CODE REVIEW

File: Multisig/Logic/Models/PrivateKey/PrivateKey.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//  Multisig
+//
+//  Created by Andrey Scherbovich on 03.09.20.
+//  Copyright © 2020 Gnosis Ltd. All rights reserved.
+//
+
+import Foundation
+import Web3
+
+
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/PrivateKey.swift:12]: Extra spaces ?

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        return Data(byteArray)
+    }
+}
+
+public class HDNode {
+    public struct HDversion{
+        public var privatePrefix: Data = Data(hex: "0x0488ADE4")
+        public var publicPrefix: Data = Data(hex: "0x0488B21E")
+        public init() {
+
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:30]: Extra spaces

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        }
+    }
+    public var path: String? = "m"
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:33]: Why "m" ?, can we use meaning and avoid constants

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+                UnsafeBufferPointer(start: $0, count: count)
+            }
+        }
+        let byteArray = Array(bytePtr)
+        return Data(byteArray)
+    }
+}
+
+public class HDNode {
+    public struct HDversion{
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:26]: Please put space before {

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        public var privatePrefix: Data = Data(hex: "0x0488ADE4")
+        public var publicPrefix: Data = Data(hex: "0x0488B21E")
+        public init() {
+
+        }
+    }
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:32]: Please add line after this initializer

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    public var path: String? = "m"
+    public var privateKey: Data? = nil
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:34]: no need to have "nil" here

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+        parentFingerprint = data[5..<9]
+        let cNum = data[9..<13].bytes
+        childNumber = UnsafePointer(cNum).withMemoryRebound(to: UInt32.self, capacity: 1) {
+            $0.pointee
+        }
+        chaincode = data[13..<45]
+        if serializePrivate {
+            privateKey = data[46..<78]
+            guard let pubKey = HDNode.privateToPublic(privateKey!, compressed: true) else {return nil}
+            if pubKey[0] != 0x02 && pubKey[0] != 0x03 {return nil}
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:88]: please put spaces before and after return nil

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+                    return nil
+                }
+                let newPK = (bn + BigUInt(self.privateKey!)) % HDNode.curveOrder
+                if newPK == BigUInt(0) {
+                    if trueIndex < UInt32.max {
+                        return self.derive(index:index+1, derivePrivateKey: derivePrivateKey, hardened:hardened)
+                    }
+                    return nil
+                }
+                guard let privKeyCandidate = newPK.serialize().setLengthLeft(32) else {return nil}
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:176]: please put space before return, also same for next line

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+                guard SECP256K1.verifyPrivateKey(privateKey: privKeyCandidate) else {return nil }
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:177]: I think they this code follow different coding style, we need to do find/replace for all {return with { return

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+            data.append(version.publicPrefix)
+        } else {
+            data.append(version.privatePrefix)
+        }
+        data.append(contentsOf: [self.depth])
+        data.append(self.parentFingerprint)
+        data.append(self.childNumber.serialize32())
+        data.append(self.chaincode)
+        if serializePublic {
+            data.append(self.publicKey)
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:293]: I don't think we need self here, same on line 296

File: Multisig/Logic/Models/PrivateKey/web3swift/Base58.swift [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+        var result: [UInt8] = Array(repeating: 0, count: zerosCount)
+        for b in base58 {
+            result.append(b)
+        }
+        return result
+    }
+}
+
+
+
```
Comment (MouazAlzahabi) [inline Multisig/Logic/Models/PrivateKey/web3swift/Base58.swift:118]: extra lines

**Review by MouazAlzahabi — APPROVED** [review]

File: Multisig/Logic/Models/PrivateKey/PrivateKey.swift (hunk 1 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/PrivateKey.swift:12]: will fix in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 2 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:30]: will fix in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 4 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:26]: will fix in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 3 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:33]: I think this is part of the standard. This code is copied, so I will not change it here.

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 5 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:32]: will do in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 6 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:34]: will do in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 7 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:88]: will do in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 8 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:176]: will do in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 9 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:177]: will do in the next PR

File: Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift (hunk 10 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/BIP32HDNode.swift:293]: I will leave it like that not to change many tiny things

File: Multisig/Logic/Models/PrivateKey/web3swift/Base58.swift (hunk 11 above)
Comment (sche) [reply Multisig/Logic/Models/PrivateKey/web3swift/Base58.swift:118]: will do in the next PR

**Pull request merged**